File: backend/app/schemas/project.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

ProjectStatus = Literal["planned", "active", "completed", "cancelled"]


class ProjectBase(BaseModel):
    client_id: int = Field(gt=0)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    status: ProjectStatus = "planned"
    start_date: date | None = None
    end_date: date | None = None
    street: str | None = Field(default=None, max_length=255)
    building_number: str | None = Field(default=None, max_length=50)
    unit_number: str | None = Field(default=None, max_length=50)
    postal_code: str | None = Field(default=None, max_length=20)
    city: str | None = Field(default=None, max_length=150)
    country_code: str = Field(default="PL", min_length=2, max_length=2)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)

    @field_validator("name", "description", "street", "building_number", "unit_number", "postal_code", "city", mode="before")
    @classmethod
    def strip_strings(cls, value: object) -> object:
        if isinstance(value, str):
            value = value.strip()
            return value or None
        return value

    @field_validator("country_code")
    @classmethod
    def normalize_country(cls, value: str) -> str:
        return value.strip().upper()

    @model_validator(mode="after")
    def validate_dates_and_coordinates(self):
        if self.end_date is not None and self.start_date is not None and self.end_date < self.start_date:
            raise ValueError("end_date cannot precede start_date")
        if (self.latitude is None) != (self.longitude is None):
            raise ValueError("latitude and longitude must be provided together")
        return self


class ProjectCreate(ProjectBase):
    pass


class ProjectUpdate(BaseModel):
    client_id: int | None = Field(default=None, gt=0)
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    status: ProjectStatus | None = None
    start_date: date | None = None
    end_date: date | None = None
    street: str | None = Field(default=None, max_length=255)
    building_number: str | None = Field(default=None, max_length=50)
    unit_number: str | None = Field(default=None, max_length=50)
    postal_code: str | None = Field(default=None, max_length=20)
    city: str | None = Field(default=None, max_length=150)
    country_code: str | None = Field(default=None, min_length=2, max_length=2)
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)


class ProjectRead(ProjectBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    client_name: str
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class ProjectPage(BaseModel):
    items: list[ProjectRead]
    total: int
    skip: int
    limit: int

File: frontend/lib/features/projects/domain/project.dart
enum ProjectStatus { planned, active, completed, cancelled }

extension ProjectStatusLabel on ProjectStatus {
  String get label => switch (this) {
    ProjectStatus.planned => 'Planowana',
    ProjectStatus.active => 'Aktywna',
    ProjectStatus.completed => 'Zakończona',
    ProjectStatus.cancelled => 'Anulowana',
  };
}

class Project {
  const Project({
    required this.id,
    required this.clientId,
    required this.clientName,
    required this.name,
    required this.status,
    required this.countryCode,
    required this.createdAt,
    required this.updatedAt,
    this.description,
    this.startDate,
    this.endDate,
    this.street,
    this.buildingNumber,
    this.unitNumber,
    this.postalCode,
    this.city,
    this.latitude,
    this.longitude,
    this.workItemId,
  });
  final int id;
  final int? workItemId;
  final int clientId;
  final String clientName;
  final String name;
  final String? description;
  final ProjectStatus status;
  final DateTime? startDate;
  final DateTime? endDate;
  final String? street;
  final String? buildingNumber;
  final String? unitNumber;
  final String? postalCode;
  final String? city;
  final String countryCode;
  final double? latitude;
  final double? longitude;
  final DateTime createdAt;
  final DateTime updatedAt;
  String get location => <String?>[
    street,
    buildingNumber,
    unitNumber,
    postalCode,
    city,
    countryCode,
  ].whereType<String>().where((value) => value.trim().isNotEmpty).join(' ');
  factory Project.fromJson(Map<String, dynamic> json) => Project(
    id: json['id'] as int,
    clientId: json['client_id'] as int,
    clientName: json['client_name']?.toString() ?? '',
    name: json['name']?.toString() ?? '',
    description: json['description']?.toString(),
    status: ProjectStatus.values.firstWhere(
      (value) => value.name == json['status'],
      orElse: () => ProjectStatus.planned,
    ),
    startDate: DateTime.tryParse(json['start_date']?.toString() ?? ''),
    endDate: DateTime.tryParse(json['end_date']?.toString() ?? ''),
    street: json['street']?.toString(),
    buildingNumber: json['building_number']?.toString(),
    unitNumber: json['unit_number']?.toString(),
    postalCode: json['postal_code']?.toString(),
    city: json['city']?.toString(),
    countryCode: json['country_code']?.toString() ?? 'PL',
    latitude: (json['latitude'] as num?)?.toDouble(),
    longitude: (json['longitude'] as num?)?.toDouble(),
    workItemId: json['work_item_id'] as int?,
    createdAt: DateTime.parse(json['created_at'].toString()),
    updatedAt: DateTime.parse(json['updated_at'].toString()),
  );
}

class ProjectPage {
  const ProjectPage({
    required this.items,
    required this.total,
    required this.skip,
    required this.limit,
  });
  final List<Project> items;
  final int total;
  final int skip;
  final int limit;
}
